near-sdk-macros/src/core_impl/code_generator/attr_sig_info.rs

```diff
@@ -191,10 +192,37 @@ impl AttrSigInfo {
                         }
                     }
                     BindgenArgType::CallbackResultArg => {
+                        let ok_type = extract_ok_type(ty).unwrap_or_else(||
+                            panic!("Function parameters marked with #[callback_result] should have \
+                                type Result<T, PromiseError>")
```

Comment:
note that this doesn't create a compiler error with span info, so the error line will be on `#[near_bindgen]` like:

```
24 | #[near_bindgen]
   | ^^^^^^^^^^^^^^^
   |
   = help: message: Function parameters marked with #[callback_result] should have type Result<T, PromiseError>
```

but you can have this error be placed on the violating line with something like https://docs.rs/syn/latest/syn/struct.Error.html and converting it into a compile error.

Can be done later, not a big deal

Reply:
nice, didn't know there were proper diagnostics in Rust macros